api/kogito-services/src/main/java/org/kie/kogito/services/uow/CollectingUnitOfWork.java

```diff
@@ -57,56 +59,78 @@ public void start() {
     @Override
     public void end() {
         checkStarted();
-        EventBatch batch = eventManager.newBatch();        
-        
-        for (WorkUnit<?> work : sorted()) {
-            batch.append(work.data());
-            work.perform();
+        EventBatch batch = eventManager.newBatch();
+
+        Optional<? extends MutableProcessInstances<?>> instances = sorted().stream()
+                .filter(work -> work instanceof ProcessInstanceWorkUnit)
+                .map(work -> ((ProcessInstanceWorkUnit<?>) work).data().process().instances())
+                .filter(ins -> ins instanceof MutableProcessInstances)
+                .map(ins -> (MutableProcessInstances<?>) ins).findAny();
+
+        if (instances.isPresent()) {
+            instances.ifPresent(i -> {
+                try (Closeable closeable = i.startTransaction()) {
+                    for (WorkUnit<?> work : sorted()) {
+                        batch.append(work.data());
+                        work.perform(closeable);
+                    }
+                    eventManager.publish(batch);
+                    i.commitTransaction(closeable);
+                    done();
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
```

Comment:
I don't think throwing a generic unchecked exception fine here. Add some logging and consider calling abortTransaction

Reply:
ok